text/0000-range-get-and-convert-methods.md

```diff
@@ -0,0 +1,397 @@
+
+Add a method to get a value from the range tied to the range.
+Later that value may be converted into another range of values,
+preserving the relative position within the new range.
+
+# Motivation
+[motivation]: #motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected outcome? -->
+
+It is helpful to have a range of possible values and be able to quickly
+obtain a value from this range. Currently, there is no way to get a value
+from a range of types `std::ops::Range` and `std::ops::RangeInclusive`;
+there is only a method called `contains()`, which can be called with a
+proposed value to check if it lies within the range. Later, if the
+value lies within the range, there is no way to tell that the value
+checked actually does that within the code: additional logic is required.
+
+A possibility for a value to be tied to a "parent" range it was got from
+will allow a value-to-new-range conversion. For example, we may want to
+have a thread priority value, which we may want to be "user-friendly" by
+having values in the range of `[0; 100]`. Later, we may pick a value out
+of this range, for example, `50`. However, on different operating systems,
+the thread priority ranges are different and depend on many things; in
+other words, it is almost certainly not the `[0; 100]` range we wanted.
+Let's assume we want to change a Linux niceness of a thread. On Linux,
+the niceness values are in the range of `[-20; 19]`. A certain calculation
+is required to map a value `50` from the range `[0; 100]` to the range
+`[-20; 19]`, to preserve the relative (middle) position, which would be
+`0` in this case (`40` allowed values in total). This can be avoided
+as these calculations can all be written once and just used. Such a
+mechanism within an already existing type like `std::ops::Range` and/or
+`std::ops::RangeInclusive` would greatly simplify this process of
+mapping values from certain ranges.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+<!-- Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means: -->
+
+By introducing a new type called `RangeValue`, we can have a type which
+declares a value tied to its parent range it was taken from:
+
+```rust
+use std::ops::Deref;
+
+/// A Range value which is tied to the range object and its lifetime.
+#[derive(Debug, Copy, Clone)]
+struct RangeValue<'r, V> {
+    value: V,
+    range: &'r std::ops::Range<V>,
+    initial: Option<Self>,
+}
+impl<'r, V> RangeValue<'r, V>
+ {
+    fn get(&self) -> &V {
+        &self.value
+    }
+
+    fn range(&self) -> &std::ops::Range<V> {
+        &self.range
+    }
+}
+
+impl<'r, V> AsRef<V> for RangeValue<'r, V> {
+    fn as_ref(&self) -> &V {
+        self.get()
+    }
+}
+
+impl<'r, V> Deref for RangeValue<'r, V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        self.get()
```

Comment:
Rust standard library types never implement `Deref` while having other methods as that could be confusing. `get` and `range` should be inherent methods without a `self` parameter.